File: projects/rpn2tex/migrations/go-module-by-module-2/parser.go
package rpn2tex

import "fmt"

// Parser converts a token stream to an Abstract Syntax Tree using
// stack-based RPN (Reverse Polish Notation) parsing algorithm.
type Parser struct {
	tokens []Token // Token stream to parse
	pos    int     // Current position in tokens (0-based)
}

// NewParser creates a new parser for the given token stream.
func NewParser(tokens []Token) *Parser {
	return &Parser{
		tokens: tokens,
		pos:    0,
	}
}

// Parse parses the token stream and returns the root AST node.
// It implements the RPN stack-based parsing algorithm:
// - Numbers are pushed onto the stack as Number nodes
// - Operators pop two operands and push a BinaryOp node
// Returns an error if the expression is invalid (empty, insufficient operands,
// or too many operands remaining on the stack).
func (p *Parser) Parse() (Expr, error) {
	var stack []Expr

	// Process all tokens except EOF
	for !p.atEnd() {
		token := p.current()

		// Skip EOF token explicitly
		if token.Type == EOF {
			break
		}

		switch token.Type {
		case NUMBER:
			// Push number node onto stack
			numNode := NewNumber(token.Line, token.Column, token.Value)
			stack = append(stack, numNode)
			p.advance()

		case PLUS, MINUS, MULT, DIV:
			// Binary operator: pop two operands, create BinaryOp node
			if len(stack) < 2 {
				return nil, NewCompileError(
					fmt.Sprintf("Operator '%s' requires two operands", token.Value),
					p.getSource(),
					token.Line,
					token.Column,
				)
			}

			// Pop right operand
			right := stack[len(stack)-1]
			stack = stack[:len(stack)-1]

			// Pop left operand
			left := stack[len(stack)-1]
			stack = stack[:len(stack)-1]

			// Map token type to operator string
			operator := p.tokenTypeToOperator(token.Type)

			// Create BinaryOp node
			opNode := NewBinaryOp(token.Line, token.Column, operator, left, right)
			stack = append(stack, opNode)
			p.advance()

		default:
			// This shouldn't happen if the lexer is correct
			return nil, NewCompileError(
				fmt.Sprintf("Unexpected token type: %s", token.Type.String()),
				p.getSource(),
				token.Line,
				token.Column,
			)
		}
	}

	// Validate final stack state
	if len(stack) == 0 {
		// Get the EOF token for error position
		eofToken := p.tokens[len(p.tokens)-1]
		return nil, NewCompileError(
			"Empty expression",
			p.getSource(),
			eofToken.Line,
			eofToken.Column,
		)
	}

	if len(stack) > 1 {
		// Too many operands - missing operators
		eofToken := p.tokens[len(p.tokens)-1]
		return nil, NewCompileError(
			fmt.Sprintf("Invalid RPN: %d values remain on stack (expected 1)", len(stack)),
			p.getSource(),
			eofToken.Line,
			eofToken.Column,
		)
	}

	return stack[0], nil
}

// atEnd returns true if the parser has reached the end of the token stream.
func (p *Parser) atEnd() bool {
	return p.pos >= len(p.tokens)
}

// current returns the current token without advancing the position.
func (p *Parser) current() Token {
	if p.atEnd() {
		// Return the last token (should be EOF)
		return p.tokens[len(p.tokens)-1]
	}
	return p.tokens[p.pos]
}

// advance moves to the next token.
func (p *Parser) advance() {
	if !p.atEnd() {
		p.pos++
	}
}

// tokenTypeToOperator converts a token type to its operator string.
func (p *Parser) tokenTypeToOperator(tokenType TokenType) string {
	switch tokenType {
	case PLUS:
		return "+"
	case MINUS:
		return "-"
	case MULT:
		return "*"
	case DIV:
		return "/"
	default:
		return ""
	}
}

// getSource reconstructs the source text from tokens for error reporting.
// This is a helper method to provide source context in error messages.
func (p *Parser) getSource() string {
	// We don't have direct access to the original source, but we can
	// reconstruct a reasonable approximation from tokens.
	// For better error reporting, the caller should pass the original source.
	// For now, return empty string and let the CompileError handle it.
	return ""
}

File: projects/rpn2tex/migrations/java-feature-by-feature-2/src/main/java/com/rpn2tex/Parser.java
package com.rpn2tex;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.List;

/**
 * Parser for RPN expressions.
 * Converts a stream of tokens into an AST using a stack-based algorithm.
 */
public class Parser {
    private final List<Token> tokens;
    private int current;

    /**
     * Creates a new parser for the given token stream.
     *
     * @param tokens List of tokens to parse
     */
    public Parser(List<Token> tokens) {
        this.tokens = tokens;
        this.current = 0;
    }

    /**
     * Parses the token stream into an expression AST.
     *
     * @return Root expression node
     * @throws RpnException If the expression is invalid
     */
    public Expr parse() throws RpnException {
        Deque<Expr> stack = new ArrayDeque<>();

        while (currentToken().type() != TokenType.EOF) {
            Token token = currentToken();

            if (token.type() == TokenType.NUMBER) {
                Number numNode = new Number(
                    token.line(),
                    token.column(),
                    token.value()
                );
                stack.push(numNode);
                advance();
            } else if (token.type() == TokenType.PLUS
                    || token.type() == TokenType.MINUS
                    || token.type() == TokenType.MULTIPLY
                    || token.type() == TokenType.DIVIDE) {
                if (stack.size() < 2) {
                    throw new RpnException(
                        "Operator '" + token.value() + "' requires two operands",
                        token
                    );
                }

                Expr right = stack.pop();
                Expr left = stack.pop();

                String operator = switch (token.type()) {
                    case PLUS -> "+";
                    case MINUS -> "-";
                    case MULTIPLY -> "*";
                    case DIVIDE -> "/";
                    default -> throw new RpnException("Unknown operator: " + token.type(), token);
                };

                BinaryOp opNode = new BinaryOp(
                    token.line(),
                    token.column(),
                    operator,
                    left,
                    right
                );
                stack.push(opNode);
                advance();
            } else {
                throw new RpnException("Unexpected token: " + token.type(), token);
            }
        }

        if (stack.isEmpty()) {
            throw new RpnException("Empty expression", currentToken());
        }

        if (stack.size() > 1) {
            throw new RpnException("Too many operands", currentToken());
        }

        return stack.pop();
    }

    private Token currentToken() {
        return tokens.get(current);
    }

    private void advance() {
        if (current < tokens.size() - 1) {
            current++;
        }
    }
}
